fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java

```diff
@@ -484,111 +576,116 @@ private static void validateRecords(@Nonnull Descriptors.FileDescriptor fileDesc
-                    case RECORD:
-                    default:
-                        break;
-                }
-                if (processExtensionOptions && recordTypeOptions.hasSinceVersion()) {
-                    sinceVersion = recordTypeOptions.getSinceVersion();
-                }
-                if (processExtensionOptions && recordTypeOptions.hasRecordTypeKey()) {
-                    recordTypeKey = LiteralKeyExpression.fromProto(recordTypeOptions.getRecordTypeKey()).getValue();
+                case UNION:
+                    if (unionHasMessageType(unionDescriptor, descriptor)) {
+                        throw new MetaDataException("Union message type " + descriptor.getName()
+                                                    + " cannot be a union field.");
+                    }
+                    continue;
+                case NESTED:
+                    if (unionHasMessageType(unionDescriptor, descriptor)) {
+                        throw new MetaDataException("Nested message type " + descriptor.getName()
+                                                    + " cannot be a union field.");
+                    }
+                    continue;
+                case RECORD:
+                    break;
+                default:
+                    break;
                 }
             }
+
             if ("RecordTypeUnion".equals(descriptor.getName())) {
-                if (union != null) {
-                    throw new MetaDataException("Only one union descriptor is allowed");
+                if (unionHasMessageType(unionDescriptor, descriptor)) {
+                    throw new MetaDataException("Union message type " + descriptor.getName()
+                                                + " cannot be a union field.");
                 }
-                union = descriptor;
-                continue;
-            }
-
-            RecordTypeBuilder recordType = new RecordTypeBuilder(descriptor);
-            recordTypes.put(recordType.getName(), recordType);
-            if (processExtensionOptions) {
-                recordType.setSinceVersion(sinceVersion);
-                recordType.setRecordTypeKey(recordTypeKey);
-                protoFieldOptions(recordType, descriptor);
+            } else if (!unionHasMessageType(unionDescriptor, descriptor)) {
+                throw new MetaDataException("Record message type " + descriptor.getName()
+                                            + " must be a union field.");
             }
         }
-        if (union == null) {
-            throw new MetaDataException("Union descriptor is required");
-        }
-        fillUnionFields(union, processExtensionOptions);
-        return union;
     }
 
+    private static boolean unionHasMessageType(@Nonnull Descriptors.Descriptor unionDescriptor, @Nonnull Descriptors.Descriptor descriptor) {
+        return unionDescriptor.getFields().stream().filter(field -> descriptor.equals(field.getMessageType())).findAny().isPresent();
```

Comment:
Sure. Suppose someone has a "union descriptor" that looks like this:

```proto
message MalformedUnion {
    optional RecordTypeOne _RecordTypeOne = 1;
    optional int32 _NotAMessageType = 2;
}
```

Suppose then one calls this method with, say, a message type called `SomeMessageType`. Then when this is called, eventually the `SomeMessageType` descriptor will be compared for equality with `MalformedUnion.getDescriptor().findFieldByName("_NotAMessageType").getMessagteType()`...and that function throws an `UnsupportedOperationException`. Which, like, the union is malformed, but the error is way too leaky. (It should throw a `MetaDataException` complaining about having non-message type fields in the union which is already something that happens in `fillUnionFields`.) I think this validation should occur before this method is invoked in order to avoid that happening.

----

Also, taking a closer look at this method it should probably use pointer equality (rather than `.equals` equality) on the descriptors, and instead of `filter`ing and checking if `findAny` returns a present `Optional`, it should probably just call `anyMatch`.

Reply:
>  I think this validation should occur before this method is invoked in order to avoid that happening.

As it stands now, this will not happen, right? As you also said, not-message-type validation happens in fillUnionFields which is called before validateUnion. At this point, we're certain that all union fields are message types. I don't mind adding the check to here too, if that's the ask? 

> Also, taking a closer look at this method it should probably use pointer equality (rather than .equals equality) on the descriptors, and instead of filtering and checking if findAny returns a present Optional, it should probably just call anyMatch.

sure.